### PR TITLE
Only compute the trait map once

### PR DESCRIPTION
**Pull request by cjgillot**
Part of #85153

r? @Aaron1011

### CODE REVIEW

**Comment by cjgillot** [comment]
Let's be prudent this time.
@bors try @rust-timer queue

File: compiler/rustc_resolve/src/lib.rs [hunk 1]
```diff
@@ -1138,8 +1140,13 @@ impl ResolverAstLowering for Resolver<'_> {
...
 
-    fn trait_map(&self) -> &NodeMap<Vec<TraitCandidate>> {
-        &self.trait_map
+    fn take_trait_map(&mut self) -> NodeMap<Vec<TraitCandidate>> {
+        #[cfg(debug_assertions)]
+        {
+            debug_assert!(!self.took_trait_map);
+            self.took_trait_map = true;
+        }
+        std::mem::take(&mut self.trait_map)
```
Comment (Aaron1011) [inline compiler/rustc_resolve/src/lib.rs:1149]: On further though - how much work would it be to make this an `Option`, and remove `took_trait_map`?

**Pull request merged**